File: server/src/services/state-persistence.ts
import fs from 'fs';
import path from 'path';
import { DB_PATH } from '../db/connection.js';
import { snapshotRateLimitState, restoreRateLimitState, type RateLimitSnapshot } from './ratelimit.js';
import { snapshotRouterState, restoreRouterState, type RouterSnapshot } from './router.js';
import { snapshotHealthState, restoreHealthState, type HealthSnapshot } from './health.js';

const STATE_FILE = path.resolve(path.dirname(DB_PATH), 'runtime-state.json');

interface FullSnapshot {
  version: 1;
  savedAt: number;
  rateLimit: RateLimitSnapshot;
  router: RouterSnapshot;
  health: HealthSnapshot;
}

export function saveRuntimeState(): void {
  const snapshot: FullSnapshot = {
    version: 1,
    savedAt: Date.now(),
    rateLimit: snapshotRateLimitState(),
    router: snapshotRouterState(),
    health: snapshotHealthState(),
  };

  const tmp = STATE_FILE + '.tmp';
  fs.writeFileSync(tmp, JSON.stringify(snapshot), 'utf-8');
  fs.renameSync(tmp, STATE_FILE);
  //console.log(`[StatePersistence] Saved runtime state to ${STATE_FILE}`);
}

export function restoreRuntimeState(): boolean {
  if (!fs.existsSync(STATE_FILE)) {
    console.log('[StatePersistence] No saved runtime state found');
    return false;
  }

  try {
    const raw = fs.readFileSync(STATE_FILE, 'utf-8');
    const snapshot: FullSnapshot = JSON.parse(raw);

    if (snapshot.version !== 1) {
      console.warn('[StatePersistence] Unknown state version, skipping restore');
      return false;
    }

    const ageMinutes = (Date.now() - snapshot.savedAt) / 60_000;
    if (ageMinutes > 60) {
      console.log(`[StatePersistence] Saved state is ${ageMinutes.toFixed(0)}min old, discarding stale state`);
      fs.unlinkSync(STATE_FILE);
      return false;
    }

    restoreRateLimitState(snapshot.rateLimit);
    restoreRouterState(snapshot.router);
    restoreHealthState(snapshot.health);

    console.log(`[StatePersistence] Restored runtime state (saved ${ageMinutes.toFixed(1)}min ago)`);
    return true;
  } catch (err) {
    console.error('[StatePersistence] Failed to restore state:', err);
    return false;
  }
}

export function clearRuntimeState(): void {
  if (fs.existsSync(STATE_FILE)) {
    fs.unlinkSync(STATE_FILE);
  }
}

let persistInterval: ReturnType<typeof setInterval> | null = null;

export function startPeriodicSave(intervalMs = 30_000): void {
  if (persistInterval) return;
  persistInterval = setInterval(() => {
    try { saveRuntimeState(); } catch (e) { console.error('[StatePersistence] Periodic save failed:', e); }
  }, intervalMs);
}

export function stopPeriodicSave(): void {
  if (persistInterval) {
    clearInterval(persistInterval);
    persistInterval = null;
  }
}

export function registerShutdownHandlers(): void {
  const shutdown = () => {
    console.log('[StatePersistence] Shutting down, saving state...');
    try { saveRuntimeState(); } catch (e) { console.error('[StatePersistence] Shutdown save failed:', e); }
    stopPeriodicSave();
    process.exit(0);
  };

  process.on('SIGTERM', shutdown);
  process.on('SIGINT', shutdown);
}

File: crates/server/src/services/state_persistence.rs
//! Persistence and restoration of in-memory runtime state.

use serde::{Deserialize, Serialize};

use crate::db::connection::db_path;
use crate::services::{health, ratelimit, router};

#[derive(Serialize, Deserialize)]
struct FullSnapshot {
    version: u32,
    #[serde(rename = "savedAt")]
    saved_at: i64,
    #[serde(rename = "rateLimit")]
    rate_limit: ratelimit::RateLimitSnapshot,
    router: router::RouterSnapshot,
    health: health::HealthSnapshot,
}

fn state_file() -> std::path::PathBuf {
    let path = db_path();
    let db = std::path::Path::new(&path);
    db.parent()
        .unwrap_or(std::path::Path::new("."))
        .join("runtime-state.json")
}

pub fn save_runtime_state() -> Result<(), String> {
    let snapshot = FullSnapshot {
        version: 1,
        saved_at: chrono::Utc::now().timestamp_millis(),
        rate_limit: ratelimit::snapshot_rate_limit_state(),
        router: router::snapshot_router_state(),
        health: health::snapshot_health_state(),
    };

    let file = state_file();
    let tmp = file.with_extension("json.tmp");
    let json = serde_json::to_string(&snapshot).map_err(|e| e.to_string())?;
    std::fs::write(&tmp, json).map_err(|e| e.to_string())?;
    std::fs::rename(&tmp, &file).map_err(|e| e.to_string())?;
    Ok(())
}

pub fn restore_runtime_state() -> bool {
    let file = state_file();
    if !file.exists() {
        tracing::info!("[StatePersistence] No saved runtime state found");
        return false;
    }

    let result: Result<bool, String> = (|| {
        let raw = std::fs::read_to_string(&file).map_err(|e| e.to_string())?;
        let snapshot: FullSnapshot = serde_json::from_str(&raw).map_err(|e| e.to_string())?;

        if snapshot.version != 1 {
            tracing::warn!("[StatePersistence] Unknown state version, skipping restore");
            return Ok(false);
        }

        let age_minutes =
            (chrono::Utc::now().timestamp_millis() - snapshot.saved_at) as f64 / 60_000.0;
        if age_minutes > 60.0 {
            tracing::info!(
                "[StatePersistence] Saved state is {}min old, discarding stale state",
                age_minutes.round()
            );
            std::fs::remove_file(&file).ok();
            return Ok(false);
        }

        ratelimit::restore_rate_limit_state(snapshot.rate_limit);
        router::restore_router_state(snapshot.router);
        health::restore_health_state(snapshot.health);

        tracing::info!(
            "[StatePersistence] Restored runtime state (saved {:.1}min ago)",
            age_minutes
        );
        Ok(true)
    })();

    match result {
        Ok(restored) => restored,
        Err(err) => {
            tracing::error!("[StatePersistence] Failed to restore state: {err}");
            false
        }
    }
}

pub fn clear_runtime_state() {
    let file = state_file();
    if file.exists() {
        std::fs::remove_file(&file).ok();
    }
}

/// `startPeriodicSave(30_000)`
pub fn start_periodic_save(interval_ms: u64) {
    static STARTED: std::sync::OnceLock<std::sync::Mutex<bool>> = std::sync::OnceLock::new();
    let started = STARTED.get_or_init(|| std::sync::Mutex::new(false));
    {
        let mut s = started.lock().unwrap();
        if *s {
            return;
        }
        *s = true;
    }
    tokio::spawn(async move {
        let mut ticker = tokio::time::interval(std::time::Duration::from_millis(interval_ms));
        ticker.tick().await; // skip immediate first tick
        loop {
            ticker.tick().await;
            if let Err(e) = save_runtime_state() {
                tracing::error!("[StatePersistence] Periodic save failed: {e}");
            }
        }
    });
}
